osu.Server.Spectator/Hubs/MultiplayerHub.cs

```diff
@@ -59,45 +47,44 @@ public async Task<MultiplayerRoom> JoinRoom(long roomId)
                 throw new InvalidStateException("Can't join a room when restricted.");
             }
 
-            var state = await GetLocalUserState();
-
-            if (state != null)
+            using (var userUsage = await GetOrCreateLocalUserState())
             {
-                // if the user already has a state, it means they are already in a room and can't join another without first leaving.
-                throw new InvalidStateException("Can't join a room when already in another room.");
-            }
+                if (userUsage.Item != null)
+                {
+                    // if the user already has a state, it means they are already in a room and can't join another without first leaving.
+                    throw new InvalidStateException("Can't join a room when already in another room.");
+                }
 
-            // add the user to the room.
-            var roomUser = new MultiplayerRoomUser(CurrentContextUserId);
+                // add the user to the room.
+                var roomUser = new MultiplayerRoomUser(CurrentContextUserId);
 
-            // check whether we are already aware of this match.
-            if (!TryGetRoom(roomId, out var room))
-            {
-                room = await RetrieveRoom(roomId);
+                MultiplayerRoom room;
 
-                room.Host = roomUser;
-
-                lock (active_rooms)
+                using (var roomUsage = await ACTIVE_ROOMS.GetForUse(roomId, true))
                 {
-                    Console.WriteLine($"Tracking new room {roomId}.");
-                    active_rooms.Add(roomId, room);
+                    if (roomUsage.Item == null)
+                    {
+                        var retrievedRoom = await RetrieveRoom(roomId);
```

Comment:
Side note: So this usage (and similar others) has a DB operation under a lock within, which makes all (non-read) DB operations touching a single room essentially serialised. Which is *proooobably* good (it's definitely easy to reason about); not sure how well it will scale, but you probably know better than I do off experience alone.

Reply:
This is intentional, and yes, it may limit performance and scalability, but I am just looking to keep things simple to try and get state consistent before adding optimisations.

For reference, bancho queues all non-essential database operations (anything which *don't* need to be executed in sequence to ensure consistency) to a separate queue class, and I intend to do the same here eventually. These lower priority operations would occur outside of the lock after retrieving the data they need from the usage.